File: src/Nullref.FullStackDemo.Frontend/src/app/api/models/widget-model.ts
/* tslint:disable */
/* eslint-disable */
import { FruitConstants } from './fruit-constants';
export interface WidgetModel {

  /**
   * The widget code
   */
  code: string;
  createdDate: string;

  /**
   * A tooltip for description
   */
  description: string;
  id: string;
  isActive: boolean;
  myFruit: FruitConstants;

  /**
   * What state is this thing from?
   */
  state: string;
}

export const WidgetModelMetadata = {
  identifier: {
    code: 'code',
    createdDate: 'createdDate',
    description: 'description',
    id: 'id',
    isActive: 'isActive',
    myFruit: 'myFruit',
    state: 'state',
  },
  maxLength: {
    code: 100,
    description: 500,
    state: 50,
  },
  minLength: {
    code: 1,
    description: 1,
    state: 1,
  },
  readOnly: {
    code: false,
    createdDate: true,
    description: false,
    id: true,
    isActive: false,
    myFruit: false,
    state: false,
  },
  required: {
    code: true,
    createdDate: true,
    description: true,
    id: true,
    isActive: true,
    myFruit: true,
    state: true,
  },
  nullable: {
    code: false,
    createdDate: false,
    description: false,
    id: false,
    isActive: false,
    myFruit: false,
    state: false,
  },
  default: {
    code: 'My defined default value',
    description: 'what....??',
    isActive: true,
    myFruit: FruitConstants.Apple,
    state: 'Georgia',
  },
  format: {
    id: 'uuid',
  },
  allowSort: {
    code: true,
    createdDate: false,
    description: false,
    id: false,
    isActive: true,
    myFruit: false,
    state: true,
  },
  displayName: {
    code: 'Code',
    createdDate: 'CreatedDate',
    description: 'My custom header',
    id: 'Id',
    isActive: 'IsActive',
    myFruit: 'MyFruit',
    state: 'State of residence',
  },
  description: {
    code: 'The widget code',
    createdDate: '',
    description: 'A tooltip for description',
    id: '',
    isActive: '',
    myFruit: '',
    state: 'What state is this thing from?',
  },
}

File: src/Nullref.FullStackDemo.API/Widget/Models/WidgetModel.cs
using Nullref.FullStackDemo.CommonModels.Attributes;
using Nullref.FullStackDemo.Database.Entity;
using static Nullref.FullStackDemo.Database.Entity.Widget;

namespace Nullref.FullStackDemo.API.Widget.Models
{
    [ReadOnly(true)]
    public class WidgetModel : WidgetUpdateModel
    {
        [Required]
        public Guid Id { get; internal set; }

        [Required]
        [LocalDateTimeRange]
        public NodaTime.LocalDateTime CreatedDate { get; internal set; } = DateTime.UtcNow.ToLocalDateTime();
    }

    public class WidgetUpdateModel : AbstractValidatableModel, IModel
    {
        [AutoTrim]
        [Searchable]
        [Sortable(IsDefault = true)]
        [Required]
        [MaxLength(MaxLengthValues.Code)]
        [Description("The widget code")]
        [DefaultValue("My defined default value")]
        public string Code { get; set; }

        [Sortable]
        [Required]
        [DefaultValue(DefaultValues.IsActive)]
        public bool IsActive { get; set; } = DefaultValues.IsActive;

        [AutoTrim]
        [Searchable]
        [Required]
        [MaxLength(MaxLengthValues.Description)]
        [Description("A tooltip for description")]
        [DefaultValue("what....??")]
        [DisplayName("My custom header")]
        public string Description { get; set; }

        [AutoTrim]
        [Searchable]
        [Sortable]
        [Required]
        [MaxLength(MaxLengthValues.State)]
        [DisplayName("State of residence")]
        [Description("What state is this thing from?")]
        [DefaultValue(DefaultValues.State)]
        public string State { get; set; } = DefaultValues.State;

        [Required]
        [DefaultValue(DefaultValues.MyFruit)]
        public FruitConstants MyFruit { get; set; } = DefaultValues.MyFruit;
    }
}
